server/signature.go

```diff
@@ -39,51 +45,75 @@ func checksumForFile(file string) (checksumHex string, err error) {
 	return checksumHex, nil
 }
 
-func signatureForFile(file string) (signatureHex string, err error) {
-
-	if privateKeyFile == "" {
-		log.Fatalf("Missing %s environment variable.", privateKeyEnv)
-	}
+func privateKey() (*rsa.PrivateKey, error) {
+	var err error
 
-	var checksum string
+	rsaPrivateKeyMu.Lock()
+	defer rsaPrivateKeyMu.Unlock()
 
-	if checksum, err = checksumForFile(file); err != nil {
-		return "", err
+	if rsaPrivateKey != nil {
+		return rsaPrivateKey, nil
 	}
 
-	var checksumHex []byte
-	if checksumHex, err = hex.DecodeString(checksum); err != nil {
-		return "", err
+	if privateKeyFile == "" {
+		log.Fatalf("Missing %s environment variable.", privateKeyEnv)
 	}
 
 	// Loading private key
 	var pb []byte
 	var fpk *os.File
 
 	if fpk, err = os.Open(privateKeyFile); err != nil {
-		return "", fmt.Errorf("Could not open private key: %q", err)
+		return nil, fmt.Errorf("Could not open private key: %q", err)
 	}
 	defer fpk.Close()
 
 	if pb, err = ioutil.ReadAll(fpk); err != nil {
-		return "", fmt.Errorf("Could not read private key: %q", err)
+		return nil, fmt.Errorf("Could not read private key: %q", err)
 	}
 
 	// Decoding PEM key.
 	pemBlock, _ := pem.Decode(pb)
 
-	var privateKey *rsa.PrivateKey
-	if privateKey, err = x509.ParsePKCS1PrivateKey(pemBlock.Bytes); err != nil {
-		return "", fmt.Errorf("Could not parse private key: %q", err)
+	rsaPrivateKey, err = x509.ParsePKCS1PrivateKey(pemBlock.Bytes)
+	if err != nil {
+		return nil, err
+	}
+
+	return rsaPrivateKey, nil
+}
+
+// Sign creates a signatures for a byte array.
+func Sign(hashedMessage []byte) ([]byte, error) {
+	pk, err := privateKey()
+	if err != nil {
+		return nil, err
 	}
 
 	// Checking message signature.
 	var signature []byte
-	if signature, err = rsa.SignPKCS1v15(rand.Reader, privateKey, crypto.SHA256, checksumHex); err != nil {
-		return "", fmt.Errorf("Could not create signature for file %s: %q", file, err)
+	if signature, err = rsa.SignPKCS1v15(rand.Reader, pk, crypto.SHA256, hashedMessage); err != nil {
```

Comment:
As the 3rd parameter depends on how `hashedMessage` is created, I recommend moving the `sha256.Sum256()` call to here, and change the method signature to `Sign(content, nonce)`.